### DISCUSSION TITLE
Using subscriptions with generated CRUD resolvers

### DISCUSSION THREAD

**Discussion by cgodkin** [MichalLytek/typegraphql-prisma#178] [Q&A]
Are there any examples of how one might use subscriptions with generated CRUD resolvers?

Thanks you very much!

**Comment by cgodkin** [comment]
Now I'm reading that Prisma 2 doesn't actually support subscriptions so no wonder this isn't in the wonderful generated CRUD resolvers.  I'll look for another solution.  Suggestions welcome!  Thanks.

**Comment by cgodkin** [comment]
Can I somehow "publish" a notification from a MiddlewareFn?  If this were possible I believe I could add the behavior I am trying for.  Thanks!

**Reply by MichalLytek** [reply]
You have to create `export const pubSub = new PubSub()` and read TypeGraphQL docs about external pubsub.
Then you can just import it and use it.

**Reply by cgodkin** [reply]
Thank you very much for the tip.  I managed to get it working.

For my future self, the `PubSub()` above is 
```
import { PubSub } from 'graphql-subscriptions';
```
and not the decorator from typegraphql.

**Reply by aidant** [reply]
I'm also looking into getting pub/sub working for the generated resolvers, have you got a working example you could share?

**Reply by cgodkin** [reply]
@aidant 
I don't have a full example, but I pasted a file plus the important parts of another into this gist to show you what I did in the server.

https://gist.github.com/cgodkin/ad9379db95d68464f4b203598bbb8030

Hope it helps.  Basically, I'm using middleware to be notified about every mutation.  For every mutation that succeeds, I publish a notification with the name of the mutation and the date.  One could do a lot more, but this meets our needs in our system.

For the client, I used apollo-client's useSubscription.

Good luck!

**Reply by aidant** [reply]
Thanks a lot, I've give this a crack tonight!

**Reply by Falco-Boehnke** [reply]
I Recommend finding out which model the mutation was called on (you can do that with the info object in your middleware) and then applying a filter to the subscription. That way you can decide wether to listen or not

**Comment by alacret** [comment]
@cgodkin worked perfectly; thanks for the help. Here I put a complete example with subscriptions, Auth0 integration, permissions, and triggers on mutations:

https://github.com/domo-dao/create-nodejs-prisma-typegrapqhl-typescript-package
